### PR TITLE
Migrate CI tests from TravisCI to CircleCI

### PR DESCRIPTION
**Pull request by brunoabinader**
Uses docker images from https://hub.docker.com/r/mbgl/ci/ and CircleCI workflows to migrate the jobs from TravisCI into CircleCI for improved test runtime and flexibility given by docker.

TODO:
- [x] Additional docker images:
  - [x] Added `linux-clang-3.9-node-4` image
  - [x] Added `linux-gcc-5-qt-4` image
  - [x] Added `linux-gcc-5-qt-5` image
  - [x] Update `base` image to include `locales` package
  - [x] Update `linux` image to include `lcov` package
  - [x] Bump docker images revision to 3
- [x] Artifacts:
  - [x] Render test results
- [x] File caching:
  - [x] ccache
  - [x] node_modules
- [x] Coveralls support - our current solution `coveralls-lcov` only supports Travis CI atm.
-   [ ] Fix logbt usage on CircleCI: https://github.com/mapbox/logbt#docker-considerations
    https://circleci.com/docs/2.0/executor-types
- [ ] AWS credentials for uploading data to s3://mapbox/mapbox-gl-native
  - [ ] Node package publishing
- [ ] Have the newly added CircleCI jobs marked as `required` in GitHub

Part of #8551.

### CODE REVIEW

File: scripts/circle_setup.sh [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+}
+
+export -f mapbox_export_mesa_library_path
+
+# Install and set up to load awscli
+pip install --user awscli
+export PATH="`python -m site --user-base`/bin:${PATH}"
+
+# Install coveralls gem
+gem install coveralls-lcov --no-rdoc --no-ri
```
Comment (kkaefer) [inline scripts/circle_setup.sh]: We should move all of these installs into the docker image so we don't have to install it every time.

File: scripts/circle_helper.sh [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+function mapbox_nanoseconds {
+    :
+}
+
+export JOBS=${CIRCLE_NODE_TOTAL:-6}
+export -f mapbox_fold
+export -f mapbox_nanoseconds
+export -f mapbox_time
+export -f mapbox_time_start
+export -f mapbox_time_finish
```
Comment (kkaefer) [inline scripts/circle_helper.sh]: Why do we need this?

File: scripts/circle_helper.sh [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+function mapbox_fold {
+    :
+}
+
+function mapbox_nanoseconds {
+    :
+}
+
+export JOBS=${CIRCLE_NODE_TOTAL:-6}
```
Comment (kkaefer) [inline scripts/circle_helper.sh]: `CIRCLE_NODE_TOTAL` refers to the concurrently running nodes per build, not to thread concurrency.

**Comment by kkaefer** [comment]
I'm skeptical about file caching, in particular for npm/mason: Circle's connectivity is excellent, and we wouldn't gain much storing these to S3.

For `ccache`: This might be viable; Circle has some built-in caching features: https://circleci.com/docs/2.0/configuration-reference/#save_cache

File: scripts/circle_helper.sh (hunk 3 above)
Comment (brunoabinader) [inline scripts/circle_helper.sh]: Yes - in that case I can just use `export JOBS=${JOBS:-6}`.

File: scripts/circle_helper.sh (hunk 2 above)
Comment (brunoabinader) [inline scripts/circle_helper.sh]: No real need in the final change - I've added this in the beginning to speed up the initial migration efforts - we can totally get rid of this 👍

File: scripts/circle_setup.sh (hunk 1 above)
Comment (brunoabinader) [inline scripts/circle_setup.sh]: It turns out these are already included (:tada:):
- `awscli`: https://github.com/mapbox/mbgl-ci-images/blob/master/base/Dockerfile#L20-L21
- `coveralls-lcov`: https://github.com/mapbox/mbgl-ci-images/blob/master/linux/Dockerfile#L12-L13

File: scripts/circle_helper.sh (hunk 3 above)
Comment (brunoabinader) [inline scripts/circle_helper.sh]: ~~Yes - in that case we can just use export JOBS=${JOBS:-6}.~~
These are not needed at all in the end - JOBS is now exported from `circle.yml`.

**Comment by friedbunny** [comment]
CircleCI’s connectivity may be top notch, but we’re already seeing issues, as in [this build](https://circleci.com/gh/mapbox/mapbox-gl-native/3232) where basic Android downloads failed:
```
> Could not resolve com.android.tools:repository:25.3.1.
  > Could not get resource 'https://jcenter.bintray.com/com/android/tools/repository/25.3.1/repository-25.3.1.pom'.
    > Could not GET 'https://jcenter.bintray.com/com/android/tools/repository/25.3.1/repository-25.3.1.pom'. Received status code 500 from server: Internal Server Error
```

**Comment by kkaefer** [comment]
@friedbunny A server error 500 indicates that the issue is with Bintray, not with Circle's connectivity? Based on the error code I believe that the same error would've happened on Bitrise/Travis as well.

**Comment by friedbunny** [comment]
@kkaefer Right, but if we’re considering caching remote packages (or building them into docker images) presumably we could partially mitigate unreliable dependency hosts?

**Comment by brunoabinader** [comment]
Machine executor type - needed for properly setting up logbt - is turning out as very tricky: You cannot run these locally, so every test attempt has to be done using CircleCI infrastructure. Also, all the steps (including checkout) are done manually inside the docker image.

We could avoid running the steps manually by using `circleci` script inside the docker images, but unfortunately `circleci` script does not provide means of using `--privileged`. We could suggest CircleCI team to add support for this, but this probably will take some time to appear. 

Another suggestion is to request CircleCI team to special-case acess to `/proc/sys/kernel/core_pattern` (needed by logbt). But again, this is something that might take some time to appear.

The affected builds are:
- `node4-clang39-release`
- `node6-clang39-release`
- `node6-clang39-debug`

I'd suggest for us to keep them running on TravisCI until we resolve this - thoughts @kkaefer @springmeyer?

**Comment by springmeyer** [comment]
> nother suggestion is to request CircleCI team to special-case acess to /proc/sys/kernel/core_pattern (needed by logbt).

Per chat I think this would be the ideal approach - since it would benefit lots of other repos and make this setup work on docker out of the box 👍

**Comment by brunoabinader** [comment]
I was not able to reproduce the error from https://circleci.com/gh/mapbox/mapbox-gl-native/3776 locally:
```
ninja: error: build.ninja:2978: expected '=', got ':'
apitrace: loaded into /usr/bin/git
        ^ near here
apitrace: unloaded from /src/platform/linux/ninja
Makefile:352: recipe for target 'node' failed
make: *** [node] Error 1
apitrace: unloaded from /usr/bin/make
[logbt] saw 'apitrace' exit with code:2 (INT)
[logbt] No corefile found at /tmp/logbt-coredumps/core.1290.*
Exited with code 2
```
Inside `build.ninja`, nothing seems to be wrong on line 2978:
```
build CMakeFiles/mbgl-core.dir/src/mbgl/util/version.cpp.o: CXX_COMPILER__mbgl-core ../../../src/mbgl/util/version.cpp
```
This makes me wonder this could be an OOM issue masked as something else - makes sense if we consider we're using logbt + apitrace together on a debug build.

**Pull request merged**